Add test design document (#408)
gardener/cert-management: docs/development/testing.md

- We use [envtest](https://pkg.go.dev/sigs.k8s.io/controller-runtime/pkg/envtest) for [integration tests](#integration-tests), as well as [pebble](https://github.com/letsencrypt/pebble) as a testing ACME Server.
- We conform to the general Gardener testing guidelines. Please read the chapter [Writing Tests (Relevant for All Kinds)](https://gardener.cloud/docs/gardener/testing/#writing-tests-relevant-for-all-kinds) in the Documentation.

## Unit Tests

### Running Unit Tests

```shell
make test
```

### Running Unit Tests with Coverage

```shell
make test-cov
```

This will create the machine-readable `./test.coverprofile` as well as a human-readable `test.coverage.html` file, which can be opened with a browser.

### Show Coverage in IDE

#### VS Code

You can show the coverage directly inside the VS Code editor in **two different ways**:

**First Way**

If not already done, you have to install the [Go Extension](https://marketplace.visualstudio.com/items?itemName=golang.Go).

On macOS, press `cmd + shift + p` (or on Windows, `ctrl + shift + p`) and select `Go: Toggle Test Coverage In Current Package`.

> You have to do this every time the code changes.

Author: You don't have to re-run `make test-cov`. VS Code runs the Tests for the package as soon as you select `Go: Toggle Test Coverage In Current Package`. It works even if there is no coverage report at all.

```
project-root/
└── pkg/
    └── cert/
        └── utils/
            ├── utils_certificate.go
            └── mock
                ├── doc.go
                └── mocks.go
```

To create the mocks, you should create a `doc.go` file inside the mock directory.
This could look like this:

```go
// SPDX-FileCopyrightText: SAP SE or an SAP affiliate company and Gardener contributors
//
// SPDX-License-Identifier: Apache-2.0

//go:generate mockgen -package=mock -destination=mocks.go github.com/gardener/controller-manager-library/pkg/resources Object

package mock
```

This uses `go:generate` to create the `mocks.go` file.
This is important, because that way the mocks automatically get (re)generated by running `make generate`.

Keep in mind to run `make generate` every time you change an interface that is mocked.

### The Purpose of Unit Tests

- Unit tests prove the correctness of a single unit according to the specification of its interface.
- Example units: functions, structs, predicates, event handlers.
- Unit tests should always be independent of one another.

### What should be covered?

- Unit tests have to cover **all important cases of input** and cover **edge cases**.

> - While test coverage can be a good tool while writing tests, it is **not required to cover all lines of code**.

Reviewer: Nit: Avoiding repetition
```suggestion
- While test coverage can be a good tool when writing tests, it is **not required to cover all lines of code**.
```

    - e.g., unit tests do not have to cover trivial functions like getter functions that only return a value.

> - Furthermore, full line coverage does not necessarily mean that you have covered all cases of valid input.

Reviewer: Why only valid input? Unit tests should also test "unhappy" paths by using invalid/problematic input on purpose

        - Please write [functional tests](#functional-tests) in these cases.

### Running Integration Tests

```shell
make test-integration
```

This will automatically set up everything you need for the integration tests.

### Set the KUBEBUILDER_ASSETS environment variable

If you want to execute or debug the integration tests **within your IDE instead of the terminal**, you have to set the `KUBEBUILDER_ASSETS` environment variable. 

Run the integration tests with the following command:

```shell
make test-integration | grep 'using envtest tools installed at'
```

This will give you the path to the local `envtest` installation:

```shell
using envtest tools installed at '/home/username/.local/share/kubebuilder-envtest/k8s/1.30.0-linux-amd64'
```

#### VS Code

You can set environment variables for Go tests inside the `settings.json` file.

Here you can add the environment variables inside the `go.testEnvVars` object.
With the example from above, the `settings.json` would look like this:

```json
{
    "go.testEnvVars": {
        "KUBEBUILDER_ASSETS": "/home/username/.local/share/kubebuilder-envtest/k8s/1.30.0-linux-amd64"
    }
}
```

> [//]: <> (TODO: Add GoLand Documentation here)

Reviewer: Similar to the coverage above I'd refer to the GoLand documentation for editing run/debug configurations:
https://www.jetbrains.com/help/go/run-debug-configuration.html

It's enough to know that you have to set the env variable correctly.

    - That way it is possible to catch bugs that can only happen when running the controller against a real Kubernetes API.

### What should be covered?

- **Larger components**, especially the `Reconcile` methods of controllers.
- **Concrete use cases**, e.g., the creation of a certificate by an issuer.
- Generally, there should be **at least one integration test for every controller**.

### What should not be covered?

- Integration tests should **not cover every detailed test case**.
- They should, however, cover the **usual positive and negative cases** that will occur during normal operation.
- Generally, small units should rather be covered extensively in the [unit tests](#unit-tests) instead of integration tests.

### Writing integration tests

Generally, the same guidelines as stated in the [Writing Integration Tests](https://gardener.cloud/docs/gardener/testing/#writing-integration-tests) section of the [Testing Strategy and Developer Guideline](https://gardener.cloud/docs/gardener/testing/) of the Gardener repository apply for the Cert-Management.

## Functional Tests

> Instead of E2E tests, we use functional tests.

Reviewer: I'd rephrase this, the functional tests are (as of today) the E2E tests of this component. It's just important to understand how they are set up and what they cover. (especially since they're being run from the make e2e target).

> Similar to E2E tests, functional tests **run in a Kind Cluster under "real world" conditions**, e.g., they do communicate with the **Staging Service of Let's Encrypt**.

Reviewer: Not everybody is familiar with KinD, let's reference it and do the same for Let's Encrypt staging 
```suggestion
Similar to E2E tests, functional tests **run in a [KinD](https://kind.sigs.k8s.io/) Cluster under "real world" conditions**, e.g., they do communicate with the [**Staging Environment of Let's Encrypt**](https://letsencrypt.org/docs/staging-environment/).

```

As opposed to E2E tests, functional tests test **single features and not the End-To-End usage**.
In the Cert-Management project, functional tests replace the classical E2E tests.

### Running Functional Tests

For the functional tests, use the test-e2e-local Make target by executing the following command:

```shell
make test-e2e-local
```

This will execute the following make targets:
- **kind-up**

>     - Create the KIND cluster needed for the functional tests.

Reviewer: Nit: The usual casing I'm familiar with is **K**ubernetes **in** **D**ocker
```suggestion
    - Create the KinD cluster needed for the functional tests.
```

- **certman-up**
    - Set-Up the Cert-Manager.
- **test-functional-local**
    - Execute the functional tests.
- **certman-dnsrecords-up**
    - Build and deploy the cert-manager with DNS records configuration.
- **test-functional-local-dnsrecords**
    - Run functional tests with DNS records.

### The Purpose of Functional Tests

- Functional tests ensure that the **functional correctness of a feature** and that the feature **conforms to its specification**.
- They are supposed to catch bugs resulting from interaction between multiple components.
- Always test from the **perspective of the user**.
- The test cases should be as close as possible to the **real usage** of the user.

### What should be covered?

- **Every feature** should have functional tests to ensure that the feature meets its specification.
- Generally, functional tests should cover the "straight-forward-cases".
- Negative cases should be covered **if they are important to the user**.
- Functional tests should cover everything that **cannot be covered with integration tests**.

>     - e.g., the revocation of a certificate because that is not supported by pebble.

Reviewer: ```suggestion
    - e.g., the revocation of a certificate because that is not supported by the Pebble ACME server.
```

### What should not be covered?

- **Small units**.
- **Details of the system**:

>     - e.g., the user also wouldn’t expect that there is a kube-apiserver deployment in the seed; they rather expect that they can talk to it no matter how it is deployed.

Reviewer: I'm not sure I understand this sentence 🤔

- [Integration Tests](#integration-tests)
  * [Running Integration Tests](#running-integration-tests)
  * [Set the KUBEBUILDER_ASSETS environment variable](#set-the-kubebuilder_assets-environment-variable)
    + [VS Code](#vs-code-1)
  * [Debugging Integration Tests](#debugging-integration-tests)
  * [The Purpose of Integration Tests](#the-purpose-of-integration-tests)
  * [What should be covered?](#what-should-be-covered-1)
  * [What should not be covered?](#what-should-not-be-covered-1)
  * [Writing integration tests](#writing-integration-tests)
- [Functional Tests](#functional-tests)
  * [Running Functional Tests](#running-functional-tests)
  * [The Purpose of Functional Tests](#the-purpose-of-functional-tests)
  * [What should be covered?](#what-should-be-covered-2)
  * [What should not be covered?](#what-should-not-be-covered-2)
  * [Writing integration tests](#writing-integration-tests-1)
- [Best Practices](#best-practices)
- [Further Information](#further-information)

<!-- tocstop -->

## General Information

- [Ginkgo](https://onsi.github.io/ginkgo/) is used as the testing framework paired with [Gomega](https://github.com/onsi/gomega) as the matcher library for all tests.

> - We use [envtest](https://pkg.go.dev/sigs.k8s.io/controller-runtime/pkg/envtest) for [integration tests](#integration-tests), as well as [Pebble](https://github.com/letsencrypt/pebble) as an ACME test server.
> 
> - We conform to the general Gardener testing guidelines. Please read the chapter [Writing Tests (Relevant for All Kinds)](https://gardener.cloud/docs/gardener/testing/#writing-tests-relevant-for-all-kinds) in the documentation.

Reviewer: Nit: Formatting
```suggestion
- We use [envtest](https://pkg.go.dev/sigs.k8s.io/controller-runtime/pkg/envtest) for [integration tests](#integration-tests), as well as [Pebble](https://github.com/letsencrypt/pebble) as an ACME test server.
- We conform to the general Gardener testing guidelines. Please read the chapter [Writing Tests (Relevant for All Kinds)](https://gardener.cloud/docs/gardener/testing/#writing-tests-relevant-for-all-kinds) in the documentation.
```

        - Please write [functional tests](#functional-tests) in these cases.

### Running Integration Tests

```shell
make test-integration
```

This will automatically set up everything you need for the integration tests.

### Set the `KUBEBUILDER_ASSETS` environment variable

If you want to execute or debug the integration tests **within your IDE instead of the terminal**, you have to set the `KUBEBUILDER_ASSETS` environment variable. 

Run the integration tests with the following command:

```shell
make test-integration | grep 'using envtest tools installed at'
```

This will give you the path to the local `envtest` installation:

```shell
using envtest tools installed at '/home/username/.local/share/kubebuilder-envtest/k8s/1.30.0-linux-amd64'
```

#### VS Code

You can set environment variables for Go tests inside the `settings.json` file.

Here you can add the environment variables inside the `go.testEnvVars` object.
With the example from above, the `settings.json` would look like this:

```json
{
    "go.testEnvVars": {
        "KUBEBUILDER_ASSETS": "/home/username/.local/share/kubebuilder-envtest/k8s/1.30.0-linux-amd64"
    }
}
```

> 
> 
> #### GoLand

Reviewer: Nit: Formatting
```suggestion

#### GoLand
```

- Generally, there should be **at least one integration test for every controller**.

### What should not be covered?

- Integration tests should **not cover every detailed test case**.
- They should, however, cover the **usual positive and negative cases** that will occur during normal operation.
- Generally, small units should rather be covered extensively in the [unit tests](#unit-tests) instead of integration tests.

### Writing integration tests

Generally, the same guidelines as stated in the [Writing Integration Tests](https://gardener.cloud/docs/gardener/testing/#writing-integration-tests) section of the [Testing Strategy and Developer Guideline](https://gardener.cloud/docs/gardener/testing/) of the Gardener repository apply for the Cert-Management.

## Functional Tests

The functional tests are the E2E tests of this component.
Functional tests **run in a [KinD](https://kind.sigs.k8s.io/) Cluster under "real world" conditions**, e.g., they communicate with the [**Staging Environment of Let's Encrypt**](https://letsencrypt.org/docs/staging-environment/) for the certificate retrieval.

### Running Functional Tests

> For the functional tests, use the test-e2e-local Make target by executing the following command:

Reviewer: Nit: Formatting
```suggestion
For the functional tests, use the `test-e2e-local` Make target by executing the following command:
```
